src/CRM/CivixBundle/Resources/views/Code/test-phpunit.php.php

```diff
@@ -0,0 +1,42 @@
+<?php
+echo "<?php\n";
+?>
+
+// Allow autoloading of PHPUnit helper classes in this extension.
+$loader = new \Composer\Autoload\ClassLoader();
+$loader->add('CRM_', __DIR__);
+$loader->add('Civi\\', __DIR__);
```

Comment:
@totten 

I know I'm harping on about PSR-4, but I don't see why we shouldn't advocate it in new extensions since it's supported. 

I could be wrong, but if we're using the Civi namespace root suggested above I think we'd end up with this directory structure to namespace mapping:

Path: `{extension.root}/tests/Civi/Test/MyCompany/MySuperExtension/Services/MyClassTest.php` 
ClassName: `Civi\Test\MyCompany\MySuperExtension\Services\MyClassTest.php`

And if we went for PSR-4:

`$loader->addPsr4('Test\MyCompany\MySuperExtension\', __DIR__);`
We could remove a few levels:

Path: `{extension.root}/tests/Services/MyClassTest.php `
Namespace: `Test\MyCompany\MySuperExtension\Services\MyClassTest.php`

Reply:
PSR-4 is great, but starting it in this context would leave the system is a rather confusing state -- because it would encourage downstream to use a mix of different file-structures *within the same extension*.

Rather, consider adding a flag to specify that an extension should use namespaces and flat directory trees, e.g.

  * When creating a module pass: `civix generate:module org.example.foo --layout=psr-4`
  * In the `info.xml`, store: `<civix><file-structure>psr-4</file-structure</civix>`
  * In subcommands like `generate:test`, `generate:upgrader`, and `generate:page`, it should create files based on that convention.

If we started out by *just* changing the `generate:test` to hard-code a PSR-4 layout, then we'd be encouraging everyone to adopt a weird hybrid layout.

This is not to say that switching to PSR-4 needs to be done all-at-once/perfectly. But it is a big change, and I wouldn't want this to become *that one quirky thing* that stems from an abortive/half-hearted design-change. The opening move in a change toward PSR-4 would need to be (a) adding the new flag and (b) using that flag in a couple of commands/templates.

Reply:
Fair point on the appropriateness of making the change here. I agree that it is probably best to avoid approaching the change as part of a big single PR where everything is changed to PSR-4, not many people are going to volunteer for that one.

Flags and conditionals here seem like a bit of over-complication. I think this is where we have a single approach which is the recommended best practice by CiviCRM and document the other options to allow people to choose what they like.